compiler/rustc_mir_transform/src/simplify.rs

```diff
@@ -292,6 +297,168 @@ pub(super) fn simplify_duplicate_switch_targets(terminator: &mut Terminator<'_>)
     }
 }
 
+const SELF_LOCAL: Local = Local::from_u32(1);
+const FIELD_ZERO: FieldIdx = FieldIdx::from_u32(0);
+
+pub(super) fn remove_dead_coroutine_switch_variants(body: &mut Body<'_>) {
+    let Some(coroutine_layout) = body.coroutine_layout_raw() else {
+        // Not a coroutine; no coroutine variants to remove.
+        return;
+    };
+
+    let bb0 = &body.basic_blocks[START_BLOCK];
+
+    let is_pinned = match body.coroutine_kind().unwrap() {
+        CoroutineKind::Desugared(CoroutineDesugaring::Gen, _) => false,
+        CoroutineKind::Desugared(CoroutineDesugaring::Async, _)
+        | CoroutineKind::Desugared(CoroutineDesugaring::AsyncGen, _)
+        | CoroutineKind::Coroutine(_) => true,
+    };
+    // This is essentially our off-brand `Underefer`. This stores the set of locals
+    // that we have determined to contain references to the coroutine discriminant.
+    // If the self type is not pinned, this is just going to be `_1`. However, if
+    // the self type is pinned, the derefer will emit statements of the form:
+    //    _x = CopyForDeref (_1.0);
+    // We'll store the local for `_x` so that we can later detect discriminant stores
+    // of the form:
+    //    Discriminant((*_x)) = ...
+    // which correspond to reachable variants of the coroutine.
+    let mut discr_locals = if is_pinned {
+        let Some(stmt) = bb0.statements.get(0) else {
+            // The coroutine body may have been turned into a single `unreachable`.
+            return;
+        };
+        // We match `CopyForDeref` (which is what gets emitted from the state transform
+        // pass), but also we match *regular* `Copy`, which is what GVN may optimize it to.
+        let StatementKind::Assign(box (
+            place,
+            Rvalue::Use(Operand::Copy(deref_place)) | Rvalue::CopyForDeref(deref_place),
+        )) = &stmt.kind
+        else {
+            panic!("The first statement of a coroutine is not a self deref");
+        };
+        let PlaceRef { local: SELF_LOCAL, projection: &[PlaceElem::Field(FIELD_ZERO, _)] } =
+            deref_place.as_ref()
+        else {
+            panic!("The first statement of a coroutine is not a self deref");
+        };
+        FxHashSet::from_iter([place.as_local().unwrap()])
+    } else {
+        FxHashSet::from_iter([SELF_LOCAL])
+    };
+
+    // The starting block of all coroutines is a switch for the coroutine variants.
+    // This is preceded by a read of the discriminant. If we don't find this, then
+    // we must have optimized away the switch, so bail.
+    let StatementKind::Assign(box (discr_place, Rvalue::Discriminant(discr_local))) =
+        &bb0.statements[if is_pinned { 1 } else { 0 }].kind
```

Comment:
If we optimize out too much, we may not even have a statement left.

Reply:
Yeah, but we already check that there's at least 1 statement in https://github.com/rust-lang/rust/pull/135471/files#diff-2b899097018634da368ec3c772523e10def01aef54b950e07c125a49e374a3e3R329. I don't know if there will ever be a case where that succeeds but this fails.